File: app/src/main/java/net/opendasharchive/openarchive/publish/PublishService.java
package net.opendasharchive.openarchive.publish;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.app.job.JobInfo;
import android.app.job.JobScheduler;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.os.Message;
import android.util.Log;

import net.opendasharchive.openarchive.db.Space;
import net.opendasharchive.openarchive.MainActivity;
import net.opendasharchive.openarchive.R;
import net.opendasharchive.openarchive.db.Collection;
import net.opendasharchive.openarchive.db.Media;
import net.opendasharchive.openarchive.db.Project;
import net.opendasharchive.openarchive.services.dropbox.DropboxSiteController;
import net.opendasharchive.openarchive.services.webdav.WebDAVSiteController;
import net.opendasharchive.openarchive.util.Prefs;

import java.io.IOException;
import java.util.Date;
import java.util.HashMap;
import java.util.List;

import androidx.annotation.Nullable;
import androidx.annotation.RequiresApi;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;

import com.orm.query.Condition;
import com.orm.query.Select;

import io.scal.secureshareui.controller.ArchiveSiteController;
import io.scal.secureshareui.controller.SiteController;
import io.scal.secureshareui.controller.SiteControllerListener;

import static io.scal.secureshareui.controller.SiteController.MESSAGE_KEY_MEDIA_ID;
import static io.scal.secureshareui.controller.SiteController.MESSAGE_KEY_PROGRESS;
import static io.scal.secureshareui.controller.SiteController.MESSAGE_KEY_STATUS;

public class PublishService extends Service implements Runnable {

    private boolean isRunning = false;
    private boolean keepUploading = true;
    private Thread mUploadThread = null;
    private    SiteController sc = null;

    @Override
    public void onCreate() {
        super.onCreate();


        if (Build.VERSION.SDK_INT >= 26)
            createNotificationChannel();

    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        if (mUploadThread == null || (!mUploadThread.isAlive()))
        {
            mUploadThread = new Thread(this);
            mUploadThread.start();
        }

        return super.onStartCommand(intent, flags, startId);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        keepUploading = false;
        if (mUploadThread != null && mUploadThread.isAlive())
            mUploadThread.interrupt();

        if (sc !=null)
            sc.cancel();
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    private boolean shouldPublish ()
    {
        if (Prefs.getUploadWifiOnly())
        {
            if ( isNetworkAvailable(true))
                return true;
        }
        else if (isNetworkAvailable(false))
        {
            return true;
        }

        //try again when there is a network
        scheduleJob(this);

        return false;
    }

    public void run ()
    {
        if (!isRunning)
            doPublish();

        stopSelf();

    }

    private synchronized boolean doPublish ()
    {
        isRunning = true;
        boolean publishing = false;

        //check if online, and connected to appropriate network type
        if (shouldPublish()) {

            publishing = true;

            //get all media items that are set into queued state
            List<Media> results = null;

            Date datePublish = new Date();

            while ((results =  Select.from(Media.class)
                    .where(Condition.prop("status").eq(Media.STATUS_QUEUED+"")).orderBy("priority DESC")
                    .list()).size() > 0 && keepUploading) {

                for (Media media: results) {

                    if (media.status != Media.STATUS_UPLOADING) {
                        media.uploadDate = datePublish;
                        media.progress = 0; //should we reset this?
                        media.status = Media.STATUS_UPLOADING;
                    }

                    try {
                        uploadMedia(media);
                        Collection coll = Collection.findById(Collection.class, media.collectionId);
                        if (coll != null) {
                            coll.uploadDate = datePublish;
                            coll.save();
                            Project proj = Project.findById(Project.class, coll.projectId);
                            if (proj != null) {
                                proj.setOpenCollectionId(-1);
                                proj.save();
                            }
                        }
                        media.save();
                    } catch (IOException ioe) {
                        Log.d(getClass().getName(), "error in uploading media: " + ioe.getMessage(), ioe);
                        media.status = Media.STATUS_QUEUED;
                        media.save();
                    }

                    if (!keepUploading)
                        break;
                }

            }

            results = Media.find(Media.class, "status = ?", Media.STATUS_DELETE_REMOTE + "");

            //iterate through them, and upload one by one
            for (Media mediaDelete : results) {

                deleteMedia(mediaDelete);
            }

        }

        isRunning = false;

        return publishing;

    }

    private void uploadMedia (Media media) throws IOException
    {

        Project project = Project.getById(media.projectId);

        if (project != null) {
            HashMap<String, String> valueMap = ArchiveSiteController.getMediaMetadata(this, media);
            media.serverUrl = project.description;
            media.status = Media.STATUS_UPLOADING;
            media.save();
            notifyMediaUpdated(media);

            Space space = null;

            if (project.spaceId != -1L)
                space = Space.findById(Space.class, project.spaceId);
            else
                space = Space.getCurrentSpace();

            if (space != null) {

                if (space.type == Space.TYPE_WEBDAV)
                    sc = SiteController.getSiteController(WebDAVSiteController.SITE_KEY, this, new UploaderListener(media), null);
                else if (space.type == Space.TYPE_INTERNET_ARCHIVE)
                    sc = SiteController.getSiteController(ArchiveSiteController.SITE_KEY, this, new UploaderListener(media), null);
                else if (space.type == Space.TYPE_DROPBOX)
                    sc = SiteController.getSiteController(DropboxSiteController.SITE_KEY, this, new UploaderListener(media), null);

                if (sc != null)
                    sc.upload(space, media, valueMap);
            }

        }
        else
        {
            media.delete();
        }

    }

    private void deleteMedia (Media media)
    {

        /**
        // if user doesn't have an account
        if(account.isAuthenticated()) {
            ArchiveSiteController siteController = (ArchiveSiteController)SiteController.getSiteController(ArchiveSiteController.SITE_KEY, this, new DeleteListener(media), null);

            if (media.getServerUrl() != null) {
                String bucketName = Uri.parse(media.getServerUrl()).getLastPathSegment();
                String fileName = ArchiveSiteController.getTitleFileName(media);
                if (fileName != null)
                    siteController.delete(account, bucketName, fileName);

                siteController.delete(account, bucketName, ArchiveSiteController.THUMBNAIL_PATH);
            }

            media.delete();
        }**/
    }

    public class UploaderListener implements SiteControllerListener {

        private Media uploadMedia;

        public UploaderListener (Media media)
        {
            uploadMedia = media;
        }

        @Override
        public void success(Message msg) {

            uploadMedia.progress = uploadMedia.contentLength;
            notifyMediaUpdated(uploadMedia);
            uploadMedia.status = Media.STATUS_UPLOADED;
            uploadMedia.save();
            notifyMediaUpdated(uploadMedia);

        }

        @Override
        public void progress(Message msg) {
            Bundle data = msg.getData();
            long contentLengthUploaded = data.getLong(SiteController.MESSAGE_KEY_PROGRESS);

            Log.d("OAPublish",uploadMedia.getId() + " uploaded: " + contentLengthUploaded + "/" + uploadMedia.contentLength );

            uploadMedia.progress = contentLengthUploaded;

            notifyMediaUpdated(uploadMedia);

        }

        @Override
        public void failure(Message msg) {
            Bundle data = msg.getData();

            String jobIdString = data.getString(SiteController.MESSAGE_KEY_JOB_ID);
            int jobId = (jobIdString != null) ? Integer.parseInt(jobIdString) : -1;

            int messageType = data.getInt(SiteController.MESSAGE_KEY_TYPE);

            int errorCode = data.getInt(SiteController.MESSAGE_KEY_CODE);
            String errorMessage = data.getString(SiteController.MESSAGE_KEY_MESSAGE);
            String error = "Error " + errorCode + ": " + errorMessage;
            //  showError(error);
            // Log.d(TAG, "upload error: " + error);

            uploadMedia.status = Media.STATUS_LOCAL;
            uploadMedia.save();

            notifyMediaUpdated(uploadMedia);

        }
    };

    public class DeleteListener implements SiteControllerListener {

        private Media deleteMedia;

        public DeleteListener (Media media)
        {
            deleteMedia = media;
        }

        @Override
        public void success(Message msg) {
            Bundle data = msg.getData();

            String jobIdString = data.getString(SiteController.MESSAGE_KEY_JOB_ID);
            int jobId = (jobIdString != null) ? Integer.parseInt(jobIdString) : -1;

            int messageType = data.getInt(SiteController.MESSAGE_KEY_TYPE);
            String result = data.getString(SiteController.MESSAGE_KEY_RESULT);
           // String resultUrl = getDetailsUrlFromResult(result);

            deleteMedia.delete();
            notifyMediaUpdated(deleteMedia);

        }

        @Override
        public void progress(Message msg) {
            Bundle data = msg.getData();

            String jobIdString = data.getString(SiteController.MESSAGE_KEY_JOB_ID);
            int jobId = (jobIdString != null) ? Integer.parseInt(jobIdString) : -1;

            int messageType = data.getInt(SiteController.MESSAGE_KEY_TYPE);

            String message = data.getString(SiteController.MESSAGE_KEY_MESSAGE);
            float progressF = data.getFloat(SiteController.MESSAGE_KEY_PROGRESS);
            //Log.d(TAG, "upload progress: " + progress);
        }

        @Override
        public void failure(Message msg) {
            Bundle data = msg.getData();

            String jobIdString = data.getString(SiteController.MESSAGE_KEY_JOB_ID);
            int jobId = (jobIdString != null) ? Integer.parseInt(jobIdString) : -1;

            int messageType = data.getInt(SiteController.MESSAGE_KEY_TYPE);

            int errorCode = data.getInt(SiteController.MESSAGE_KEY_CODE);
            String errorMessage = data.getString(SiteController.MESSAGE_KEY_MESSAGE);
            String error = "Error " + errorCode + ": " + errorMessage;
            //  showError(error);
            // Log.d(TAG, "upload error: " + error);


            notifyMediaUpdated(deleteMedia);

        }
    };



    private boolean isNetworkAvailable(boolean requireWifi) {
        ConnectivityManager manager =
                (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = manager.getActiveNetworkInfo();
        boolean isAvailable = false;
        if (networkInfo != null && networkInfo.isConnected()) {
            // Network is present and connected
            isAvailable = true;

            boolean isWiFi = networkInfo.getType() == ConnectivityManager.TYPE_WIFI;
            if (requireWifi && (!isWiFi))
                return false;

        }
        return isAvailable;
    }

    // Send an Intent with an action named "custom-event-name". The Intent sent should
// be received by the ReceiverActivity.
    private void notifyMediaUpdated(Media media) {
        Log.d("sender", "Broadcasting message");
        Intent intent = new Intent(MainActivity.INTENT_FILTER_NAME);
        // You can also include some extra data.
        intent.putExtra(MESSAGE_KEY_MEDIA_ID, media.getId());
        intent.putExtra(MESSAGE_KEY_STATUS,media.status);
        intent.putExtra(MESSAGE_KEY_PROGRESS,media.progress);
        LocalBroadcastManager.getInstance(this).sendBroadcast(intent);
    }

    public static final int MY_BACKGROUND_JOB = 0;

    public static void scheduleJob(Context context) {

        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.LOLLIPOP) {

            JobScheduler js = (JobScheduler) context.getSystemService(Context.JOB_SCHEDULER_SERVICE);

            JobInfo job = new JobInfo.Builder(
                    MY_BACKGROUND_JOB,
                    new ComponentName(context, PublishJobService.class))
                    .setRequiredNetworkType(JobInfo.NETWORK_TYPE_UNMETERED)
                    .setRequiresCharging(false)
                    .build();
            js.schedule(job);
        }
    }

    private final static String NOTIFICATION_CHANNEL_ID = "oasave_channel_1";

    @RequiresApi(api = Build.VERSION_CODES.O)
    private void createNotificationChannel ()
    {
        NotificationManager mNotificationManager =
                (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
// The id of the channel

// The user-visible name of the channel.
        CharSequence name = getString(R.string.app_name);
// The user-visible description of the channel.
        String description = getString(R.string.app_subtext);
        int importance = NotificationManager.IMPORTANCE_LOW;
        NotificationChannel mChannel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, name, importance);
// Configure the notification channel.
        mChannel.setDescription(description);
        mChannel.enableLights(false);
        mChannel.enableVibration(false);
        mChannel.setShowBadge(false);
        mChannel.setLockscreenVisibility(Notification.VISIBILITY_SECRET);
        mNotificationManager.createNotificationChannel(mChannel);
    }


}

File: app/src/main/java/net/opendasharchive/openarchive/publish/PublishService.kt
package net.opendasharchive.openarchive.publish

import android.app.*
import android.app.job.JobInfo
import android.app.job.JobScheduler
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import android.os.IBinder
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import kotlinx.coroutines.runBlocking
import net.opendasharchive.openarchive.CleanInsightsManager
import net.opendasharchive.openarchive.features.main.MainActivity
import net.opendasharchive.openarchive.R
import net.opendasharchive.openarchive.db.Media
import net.opendasharchive.openarchive.services.Conduit
import net.opendasharchive.openarchive.util.Prefs
import timber.log.Timber
import java.io.IOException
import java.util.*

class PublishService : Service(), Runnable {
    private var mRunning = false
    private var mKeepUploading = true
    private var mUploadThread: Thread? = null
    private val mConduits = ArrayList<Conduit?>()

    override fun onCreate() {
        super.onCreate()

        if (Build.VERSION.SDK_INT >= 26) createNotificationChannel()

        doForeground()
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (mUploadThread?.isAlive != true) {
            mUploadThread = Thread(this)
            mUploadThread?.start()
        }

        return super.onStartCommand(intent, flags, startId)
    }

    override fun onDestroy() {
        super.onDestroy()

        mKeepUploading = false

        for (conduit in mConduits) conduit?.cancel()

        mConduits.clear()
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    private fun shouldPublish(): Boolean {
        if (Prefs.uploadWifiOnly) {
            if (isNetworkAvailable(true)) return true
        }
        else if (isNetworkAvailable(false)) {
            return true
        }

        // Try again when there is a network.
        scheduleJob(this)

        return false
    }

    override fun run() {
        if (!mRunning) doPublish()

        stopSelf()
    }

    private fun doPublish(): Boolean {
        mRunning = true

        var publishing = false

        //check if online, and connected to appropriate network type
        if (shouldPublish()) {
            publishing = true

            // Get all media items that are set into queued state.
            var results = emptyList<Media>()
            val datePublish = Date()

            while (mKeepUploading &&
                Media.getByStatus(listOf(Media.Status.Queued, Media.Status.Uploading), Media.ORDER_PRIORITY)
                    .also { results = it }.isNotEmpty()
            ) {
                for (media in results) {
                    val project = media.project

                    if (media.sStatus != Media.Status.Uploading) {
                        media.uploadDate = datePublish
                        media.progress = 0 // Should we reset this?
                        media.sStatus = Media.Status.Uploading
                        media.statusMessage = ""
                    }

                    media.licenseUrl = project?.licenseUrl

                    try {
                        if (uploadMedia(media)) {
                            val collection = media.collection
                            collection?.uploadDate = datePublish
                            collection?.save()

                            media.save()
                        }
                    }
                    catch (ioe: IOException) {
                        Timber.d(ioe)

                        media.statusMessage = "error in uploading media: " + ioe.message
                        media.sStatus = Media.Status.Error
                        media.save()
                    }

                    if (!mKeepUploading) return false // Time to end this.
                }
            }

            results = Media.getByStatus(listOf(Media.Status.DeleteRemote))

            //iterate through them, and upload one by one
            for (mediaDelete in results) {
                deleteMedia()
            }
        }

        mRunning = false

        return publishing
    }

    @Throws(IOException::class)
    private fun uploadMedia(media: Media): Boolean {
        val serverUrl = media.project?.description

        if (serverUrl.isNullOrEmpty()) {
            media.delete()

            return false
        }

        media.serverUrl = serverUrl
        media.sStatus = Media.Status.Uploading
        media.save()
        UploaderListener.notifyMediaUpdated(media, this)

        val conduit = Conduit.get(media, this, UploaderListener(media, this.applicationContext), null)
            ?: return false

        CleanInsightsManager.measureEvent("upload", "try_upload", media.space?.tType?.friendlyName)

        mConduits.add(conduit)
        runBlocking {
            conduit.upload()
        }
        mConduits.remove(conduit)

        return true
    }

    private fun deleteMedia() {

    }

    private fun isNetworkAvailable(requireWifi: Boolean): Boolean {
        val cm = getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager ?: return false

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            val cap = cm.getNetworkCapabilities(cm.activeNetwork) ?: return false

            when {
                cap.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> {
                    return true
                }
                cap.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> {
                    return !requireWifi
                }
                cap.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> {
                    return !requireWifi
                }
            }

            return false
        }
        else {
            @Suppress("DEPRECATION")
            val info = cm.activeNetworkInfo

            @Suppress("DEPRECATION")
            return info?.isConnected == true && (!requireWifi || info.type == ConnectivityManager.TYPE_WIFI)
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private fun createNotificationChannel() {
        val channel = NotificationChannel(NOTIFICATION_CHANNEL_ID, getString(R.string.app_name),
            NotificationManager.IMPORTANCE_LOW)

        channel.description = getString(R.string.app_subtext)
        channel.enableLights(false)
        channel.enableVibration(false)
        channel.setShowBadge(false)
        channel.lockscreenVisibility = Notification.VISIBILITY_SECRET

        (getSystemService(NOTIFICATION_SERVICE) as? NotificationManager)
            ?.createNotificationChannel(channel)
    }

    @Synchronized
    private fun doForeground() {
        val flag = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PendingIntent.FLAG_IMMUTABLE
        }
        else 0

        val pendingIntent = PendingIntent.getActivity(this, 0,
            Intent(this, MainActivity::class.java), flag)

        val notification = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_app_notify)
            .setContentTitle(getString(R.string.app_name))
            .setDefaults(Notification.DEFAULT_LIGHTS)
            .setContentIntent(pendingIntent)
            .build()

        startForeground(1337, notification)
    }

    companion object {
        private const val MY_BACKGROUND_JOB = 0
        private const val NOTIFICATION_CHANNEL_ID = "oasave_channel_1"

        fun scheduleJob(context: Context) {
            val job = JobInfo.Builder(MY_BACKGROUND_JOB,
                ComponentName(context, PublishJobService::class.java))
                .setRequiredNetworkType(JobInfo.NETWORK_TYPE_UNMETERED)
                .setRequiresCharging(false)
                .build()

            (context.getSystemService(JOB_SCHEDULER_SERVICE) as? JobScheduler)
                ?.schedule(job)
        }
    }
}